### ISSUE TITLE
Use of popen2 deprecated in py2.6

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3431]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#3431|
|Type|enhancement|
|Created|2008-09-12 00:50:55Z|
|Branch|https://github.com/twisted/twisted/tree/remove-popen2-3431-2|

Trial reports ```DeprecationWarnings``` for [popen2](http://docs.python.org/dev/library/popen2.html) in [twisted/test/test_process.py](twisted/test/test_process.py). [Deprecated](http://docs.python.org/dev/whatsnew/2.6.html) in py2.6.

```
DeprecationWarning: The popen2 module is deprecated.  Use the subprocess module.
```

Attachments:

* [test-process-warning.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dc6/dc6e4a95c12a8a4cf3d90f3a3a5fb3cd384252ef/f7466e925e74019da2e13238df4a57e0cd5364e0.patch) (1978 bytes) - added by gas on 2010-01-13 15:05:51Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3431 3431
type__enhancement enhancement
reporter__thijs thijs
priority__normal normal
milestone__Python_2_6 Python-2.6
branch__branches_remove_popen2_3431_2 branches/remove-popen2-3431-2
branch_author__kelly__exarkun kelly, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1221180655000000 1221180655000000
changetime__1274928962000000 1274928962000000
version__None None
owner__ 
cc__thijs cc__khorn cc__kelly
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

We could drop this entirely in favor of running a program with known output.  Of course, knowing the stderr of a Python program is pretty hard... Or maybe impossible.  Maybe it could be /bin/sh, though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @thijstriemstra|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

[#3986](https://github.com/twisted/twisted/issues/3986) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="Markon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|Markon commented|
|-|-|

Uhm, I've edited this line:

```
lsOut = popen2.popen3("/bin/ls ZZXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX")[2].read()
```

to:
```
    lsOut = subprocess.Popen("/bin/ls ZZXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
                             shell=True, stderr=subprocess.PIPE).communicate()[1]
```

And tried to run the tests using `trial test_process.py`. It passes all the tests (except SKIPPED tests). But I've used Python 2.5. 
Can someone test this new line with Py 2.6? ( you've to just add `import subprocess`, and go to the lines >= 2280, where you can see `lsOut`)
Bye!

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_7"></a>|@khorn commented|
|-|-|

this is a duplicate of [#3900](https://github.com/twisted/twisted/issues/3900)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra set owner to @khorn|
|-|-|

Un-assigning.

**Comment by twisted-trac** [comment]
|<img alt="gas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|gas removed owner|
|-|-|

Have copied over kelly patch from [#3900](https://github.com/twisted/twisted/issues/3900). It appears to work running trial in no extra errors running trial against test_process.py

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Thanks.  Note that the keywords field is case sensitive, unfortunately.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [27975]) Branching to 'remove-popen2-3431'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

I guess the branch is ready for review so there we go.

 - I know it's for internal use but ```twisted/python/_release.py``` (and it's tests) still uses popen2, bonus points for fixing that one as well..
 
Modified files need to have an updated copyright header (2010).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

The branch has another issue as well.  Somehow it fails a test in another module:

```
===============================================================================
[FAIL]: twisted.test.test_tcp_internals.PlatformAssumptionsTestCase.test_acceptOutOfFiles

Traceback (most recent call last):
  File "/var/lib/buildbot/bigdog24-twisted/ubuntu64-py2.5-select/Twisted/twisted/test/test_tcp_internals.py", line 97, in test_acceptOutOfFiles
    self.assertIn(exc.args[0], (EMFILE, ENOBUFS))
twisted.trial.unittest.FailTest: 9 not in (24, 105)
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

The test failure comes from a pipe leak in some versions of pygtk.  It's only related to the change in this branch in that somehow this change disturbs things slightly and causes the existing problem to become visible.

There _is_ a problem with `test_acceptOutOfFiles` that should be fixed - it shouldn't close its port if it can't create any other sockets - but I think that will just re-mask the underlying problem.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

I replaced the other usage of `popen2` as well.  However, merging this would still turn 3 supported builders red.

Arguably they shouldn't really be supported - they have a version of pygtk which leaks file descriptors.  What we can do about this, though, isn't obvious to me.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra commented|
|-|-|

[#4294](https://github.com/twisted/twisted/issues/4294) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra removed owner|
|-|-|

Replying to [exarkun](#note_16):
> I replaced the other usage of `popen2` as well.  However, merging this would still turn 3 supported builders red.
>
> Arguably they shouldn't really be supported - they have a version of pygtk which leaks file descriptors.  What we can do about this, though, isn't obvious to me.

Let's contact the maintainers of those buildslaves so they can upgrade their version of pygtk? I think those issues shouldn't block this ticket though and I'm putting it back up for review.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|TimAllen set owner to @exarkun|
|-|-|

Looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [29042]) Branching to 'remove-popen2-3431-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [29075]) Raise the artificial limit to 512 to work-around the pygtk fd leak

refs [#3431](https://github.com/twisted/twisted/issues/3431)

See also https://bugs.launchpad.net/ubuntu/+source/gtk+2.0/+bug/363245

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun removed owner|
|-|-|

Made a few final changes.  Nothing really huge.  Would be nice for someone to take one last look, though.  Most recent [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-popen2-3431-2).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph set owner to @exarkun|
|-|-|

Just a few nits to pick, then merge:

    1. `testStderr` should be renamed `test_stderr` as long as you're touching it.
    1. "produces the correct result", really?  `testStderr`'s docstring should phrase that in a way that doesn't leave the "correct result" up to the imagination.

Other than that, looks good, merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [29077]) Fix `testStderr` a bit

refs [#3431](https://github.com/twisted/twisted/issues/3431)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set status to `closed`|
|-|-|

(In [29078]) Merge remove-popen2-3431-2

Author: kelly, exarkun
Reviewer: thijs, TimAllen, glyph
Fixes: [#3431](https://github.com/twisted/twisted/issues/3431)

Remove the uses of the deprecated `popen2` module, replacing them with
the `subprocess` module.

Also, raise the fd limit set by a test in `test_tcp_internals`.  This is
really an unrelated change, but the `popen2` changes in this branch
shuffle things around just enough to make a separate existing issue more
likely to arise.  Raising this limit gives us some more time to fix the
real cause of these other failures (pygtk file descriptor leak, see
https://bugs.launchpad.net/ubuntu/+source/gtk+2.0/+bug/363245 for further
details).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
